sdk/compiler/damlc/daml-stdlib-src/DA/Crypto/Text.daml

```diff
@@ -78,14 +87,106 @@ class HasToHex a where
   -- | Converts a typed data value into a hex encoded string.
   toHex: a -> BytesHex
 
+instance HasToHex Text where
+  toHex = primitive @"BEEncodeHex"
+
+-- | HIDE
+intToHexByte: Int -> BytesHex
+intToHexByte n | 0 <= n && n <= 9 = show n
+intToHexByte 10 = "a"
+intToHexByte 11 = "b"
+intToHexByte 12 = "c"
+intToHexByte 13 = "d"
+intToHexByte 14 = "e"
+intToHexByte 15 = "f"
+intToHexByte n | n < 256 = (intToHexByte(n / 16)) <> (intToHexByte(n % 16))
+intToHexByte n = error $ "Int should be in range [0, 256): " <> show n
+
+instance HasToHex Int where
+  toHex n | 0 <= n && n <= 9 = "0" <> (show n)
+  toHex 10 = "0a"
+  toHex 11 = "0b"
+  toHex 12 = "0c"
+  toHex 13 = "0d"
+  toHex 14 = "0e"
+  toHex 15 = "0f"
+  toHex n | n < 256 = intToHexByte n
+  toHex n = (toHex(n / 256)) <> (toHex(n % 256))
+
+instance HasToHex Party where
+  toHex p = toHex(partyToText p)
+
+instance NumericScale n => HasToHex (Numeric n) where
+  toHex num = toHex(show num)
+
 class HasFromHex a where
   -- | Converts a hex encoded string into a typed data value.
   fromHex: BytesHex -> a
 
-instance HasToHex Text where
-  toHex = primitive @"BEEncodeHex"
-
 instance HasFromHex Text where
   fromHex "" = ""
-  fromHex arg | Text.isHex arg = primitive @"BEDecodeHex" (Text.asciiToLower arg)
-  fromHex arg = error $ "Argument is not a hex string: " <> arg
+  fromHex hex | Text.isHex hex = primitive @"BEDecodeHex" (Text.asciiToLower hex)
+  fromHex arg = error $ "Expected a hex encoded string but argument was: " <> arg
+
+instance HasFromHex Int where
+  fromHex "" = 0
+  fromHex hex | Text.isHex hex && Text.length hex == 2 = hexByteToInt hex
+  fromHex hex | Text.isHex hex =
+    let right = fromHex(Text.substring ((Text.length hex) - 2) 2 hex)
+        left = fromHex(Text.substring 0 ((Text.length hex) - 2) hex)
+    in
+      left * 256 + right
+  fromHex arg = error $ "Expected a hex encoded string but argument was: " <> arg
+
+instance HasFromHex Party where
+  fromHex hex | Text.isHex hex = fromSome(partyFromText(fromHex hex))
+  fromHex arg = error $ "Expected a hex encoded string but argument was: " <> arg
```

Comment:
I'm not convinced about pretending that `fromHex` is a total function when it really is not. Now as long as the Daml code doesn't need to recover from the error this is fine.

In particular, if all the primitives used are exposed, so that users can write their own parsers.

So fine for a 3.3 release.

Reply:
I'll refactor so that the `HasFromHex a` instances are defined for `HasFromHex (Optional a)` instances instead